### PR TITLE
feat(builder): add inline chunk plugin

### PR DESCRIPTION
**Pull request by jkzing**
# PR Details

- add PluginInlineChunk
  - add webpack plugin InlineChunkHtmlPlugin into builder
  - register InlineChunkHtmlPlugin in PluginInlineChunk and enable inline base on options in `output`
- change PluginEntry
  - add entries defined by source.vendorEntry into webpack-chain entry
- change PluginHtml
  - filter out vendorEntry before register html plugin, because they don't need a html file
  - add all vendorEntry to existing html plugin chunks option
- change PluginSplitChunk
  - set runtimeChunk to `single` mode (with name `builder-runtime`) because there could be more than one entries in a html

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] I have updated changeset
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/builder/webpack-builder/src/plugins/entry.ts [hunk 1]
```diff
@@ -10,18 +11,22 @@ export const PluginEntry = (): BuilderPlugin => ({
...
-        ensureArray(entry[entryName]).forEach(item => {
-          chain.entry(entryName).add(item as string);
+          ensureArray(entryObject[entryName]).forEach(item => {
+            chain.entry(entryName).add(item as string);
+          });
         });
-      });
+      };
+      addEntries(entry);
+      addEntries(vendorEntry);
```
Comment (chenjiahan) [inline packages/builder/webpack-builder/src/plugins/entry.ts:29]: vendorEntry 和 source.preEntry 好像挺类似的：https://modernjs.dev/docs/apis/app/config/source/pre-entry

如果把 preEntry 的内容通过 splitChunk 抽取出来，应该就和 vendorEntry 很接近了。

File: packages/builder/webpack-builder/src/plugins/splitChunks.ts [hunk 2]
```diff
@@ -215,7 +215,9 @@ export function PluginSplitChunks(): BuilderPlugin {
           userDefinedCacheGroups,
           builderConfig: chunkSplit,
         });
-        chain.optimization.splitChunks(splitChunksOptions);
+        chain.optimization.splitChunks(splitChunksOptions).runtimeChunk({
+          name: RUNTIME_CHUNK_NAME,
```
Comment (chenjiahan) [inline packages/builder/webpack-builder/src/plugins/splitChunks.ts:219]: vendorEntry 是增加独立 entry 的形式来实现的，这样会使一个 HTML 文件中包含多个 entry chunk，会有 common module 被重复执行的现象，把 runtimeChunk 设置为 single 才能避免该问题

> Imported modules are initialized for each runtime chunk separately, so if you include multiple entry points on a page, beware of this behavior. You will probably want to set it to single or use another configuration that allows you to only have one runtime instance.

https://webpack.js.org/configuration/optimization/#optimizationruntimechunk

Comment (chenjiahan) [reply packages/builder/webpack-builder/src/plugins/splitChunks.ts:219]: 看了下设置成统一的 name 应该也是等价的，那没事了~

File: packages/builder/webpack-builder/src/plugins/entry.ts (hunk 1 above)
Comment (jkzing) [reply packages/builder/webpack-builder/src/plugins/entry.ts:29]: 效果上确实差不多，但是似乎用 preEntry 也不能避免增加一堆额外的配置（比如指定哪些 preEntry 要被独立 split）🤔

Comment (chenjiahan) [reply packages/builder/webpack-builder/src/plugins/entry.ts:29]: 额外的配置可以考虑在内部自动处理掉，比如为 preEntry 设置一个 split 规则，超过一定体积就拆分为独立的 chunk，这样使用者只需要配置 preEntry 就好

Comment (chenjiahan) [inline packages/builder/webpack-builder/src/plugins/entry.ts:29]: 增加额外的 entry 还可能引入一个问题，就是构建 SSR Bundle 的时候无法被打包到对应的 Bundle 里（SSR 是每个 Entry 对应一个 bundle，不做拆包），导致 vendorEntry 的逻辑在 SSR 阶段不会执行。

Comment (jkzing) [reply packages/builder/webpack-builder/src/plugins/entry.ts:29]: 确实，在 SSR 下不太行

Comment (jkzing) [reply packages/builder/webpack-builder/src/plugins/entry.ts:29]: 用 preEntry 也还是很别扭，相当于先 bundle 在一起再 split，而原始的需求只是需要一个独立的 bundle 而已，这样等于绕了一圈来实现这个功能

Comment (jkzing) [reply packages/builder/webpack-builder/src/plugins/entry.ts:29]: 我觉得在想清楚之前，还是宁愿先不要这个功能吧...

**Comment by jkzing** [comment]
把 vendorEntry 相关内容独立到 #1688 中，这个 PR 只保留 inline chunk 的功能

**Review by chenjiahan — APPROVED** [review]

**Comment by chenjiahan** [comment]
LGTM~

**Pull request merged**